nm: Refactor NmActiveConnection and NmConnection search code

cathay4t opened the pull request:
Introducing `NmConnectionMatcher` to unify the effort on find the matching `NmActiveConnection` and `NmConnection`.

Resolves: https://issues.redhat.com/browse/RHEL-92916
Resolves: https://issues.redhat.com/browse/RHEL-86240

ihuguet commented on rust/src/lib/nm/query_apply/profile_search.rs:
```diff
@@ -0,0 +1,357 @@
+//
+// Using String instead of &str for interface name indexing because VLAN
+// connection can have auto-generated `connection.interface-name` when it is
+// empty.
```
I would remove this comment, too low level detail for such a long comment.

cathay4t replied:
I wrote it in case you ask me why store cloned `String` instead of `&str`.

ihuguet replied:
Haha so you are starting to anticipate to my reviews! So good.

ihuguet commented on rust/src/lib/nm/query_apply/profile_search.rs:
```diff
@@ -0,0 +1,357 @@
+
+    // (mac_upper_case, nm_iface_type) to Vec<Rc<NmConnection>> for mac
+    // identifier Veth will also stored into `NmIfaceType::Ethernet`
+    nm_saved_conns_mac: HashMap<(String, NmIfaceType), Vec<Rc<NmConnection>>>,
```
Instead of these comments, better to explain the content in the field name itself. Example: `active_conns_by_uuid` instead of `nm_acs`, `applied_cons_by_iface` instead of `nm_applied_conns_index`, etc.

cathay4t replied:
Changed.

ihuguet commented on rust/src/lib/nm/query_apply/profile_search.rs:
```diff
@@ -0,0 +1,357 @@
+    // identifier Veth will also stored into `NmIfaceType::Ethernet`
+    nm_saved_conns_mac: HashMap<(String, NmIfaceType), Vec<Rc<NmConnection>>>,
+
+    cur_state: NetworkState,
```
What about not storing the NetworkState in the struct and, instead, pass a `&NetworkState` as argument to those methods that need it? Advantages:
1. Avoid a clone of the NetworkState
2. If the NetworkState evolves after the creation of the SearchProfile, we can call add_nm_applied/saved_conn with the up to date NetworkState.

cathay4t replied:
As struct holding reference of external is complex. Even I managed to get it done, it will not be new-comer friendly.

cathay4t replied:
Never mind. I don't need to cache current state at all.

ihuguet replied:
Exactly, that's what I meant.

ihuguet commented on rust/src/lib/nm/query_apply/profile_search.rs:
```diff
@@ -0,0 +1,357 @@
+            log::debug!(
+                "Ignoring applied NmConnection which cannot resolve its \
+                 network interface {nm_conn:?}"
+            );
```
If it's something that must not happen, better to return an error or panic via assertion.

cathay4t replied:
User might have existing NmConnection pointing to deleted NIC, it is valid environment, we should not raise any error or warning on this. Treat is debug log can be collected when failure happens.

ihuguet replied:
AFAIK that's not possible as this is in add_nm_applied_con. The caller is supposed to add only applied connections here. All applied connections should have iface_name set, as we do in `NmApi::applied_connections_get`.
If we don't ensure that this is true in the caller, it is a logic error that we need to fix.

cathay4t replied:
I don't want to raise error for unsupported NmConnection even it is applied connection. I am bumping it to a warning message as it most likely not impact on current `nmstatectl apply` action. Failing the apply action for unrelated NmConnection is not a good idea, if there is bug here, the verification process will notice.

ihuguet replied:
Makes sense. Would you consider raising it to "error" and with the prefix "Bug:"? If this happens, it's because there is a bug in nmstate code (or maybe NM), and this way we'll notice it in an easier way. Warnings happen in a daily manner to users, and the message is not descriptive enough for them to understand what happens.

cathay4t replied:
Changed to error log with `BUG:` prefix.

ihuguet commented on rust/src/lib/nm/query_apply/profile_search.rs:
```diff
@@ -0,0 +1,357 @@
+                .and_modify(|nm_conns| nm_conns.push(nm_conn.clone()))
+                .or_insert(vec![nm_conn.clone()]);
+        }
+        if nm_conn.iface_name().unwrap_or_default().is_empty() {
```
Do we want to store by MAC only if interface name is empty? Doesn't make sense to store it anyway?

cathay4t replied:
When user have disconnected NIC, nmstate should support deleting its mac-identifier NmConnection.

ihuguet replied:
Can you expand your explanation a bit? I still don't understand why should not we save by name AND by mac.

cathay4t replied:
1. Create a down profile for MAC `00:23:45:67:89:1a` which actually does not exists.
```bash
sudo nmcli c add type ethernet connection.id wlan0  \
    802-3-ethernet.mac-address 00:ff:ff:ff:ff:1a \
    ipv4.method disable ipv6.method disable
```
2. Desire to remove this config:
```yml
---
interfaces:
  - name: wan0
    type: ethernet
    state: absent
    identifier: mac-address
    mac-address: 00:ff:ff:ff:ff:1a
```

In this case, there is not real NIC name for this `00:23:45:67:89:1a`, the name `wan0` is actually stored as `connection.id` not the `connection.interface-name` or resolvable to real kernel NIC name. So we have to use `mac` + `NmIfaceType` as key of this map index.

cathay4t replied:
I have included extra patch b8cd20fb7a2fc899654c326977d48c02bc956ae9 to explain the use of this.

cathay4t replied:
Patch expanded with new test case with both interface name and mac defined in NmConnection.

ihuguet commented on rust/src/lib/nm/query_apply/profile_search.rs:
```diff
@@ -0,0 +1,357 @@
+                self.nm_saved_conns_index
+                    .entry((name.to_string(), NmIfaceType::Ethernet))
+                    .and_modify(|nm_conns| nm_conns.push(nm_conn.clone()))
+                    .or_insert(vec![nm_conn.clone()]);
```
nit: use `or_insert_with` to avoid eagerly evaluation of the argument. With `or_insert` you will always allocate a Vec and clone the Rc.
Alternatively, I think that `.entry(...).or_insert_with(|| Vec::new()).push(nm_conn.clone())` would work too.
Same for the other 3 times in the function.

ihuguet replied:
Please consider this change request to avoid unnecessary clones and allocations.

cathay4t replied:
Fixed. Thanks.

cathay4t replied:
Based on suggestion of `cargo clippy`, changed to 

```rust
                .entry((name, nm_iface_type.clone()))
                .or_default()
                .push(nm_conn.clone())
```

ihuguet commented on rust/src/lib/nm/profile_search.rs:
```diff
@@ -0,0 +1,367 @@
+        self.nm_acs.contains_key(uuid)
+    }
+
+    pub(crate) fn get_applied_by_name_type(
```
Suggestion: get_applied_by_iface

cathay4t replied:
We already have `get_applied()` which takes `BaseInterface`, so this `by_iface` might be not fit here.

ihuguet commented on rust/src/lib/nm/profile_search.rs:
```diff
@@ -0,0 +1,367 @@
+        return Some(format!("{vlan_parent}.{vlan_id}"));
+    }
+
+    if let Some(mac) = nm_conn
+        .wired
+        .as_ref()
+        .and_then(|s| s.mac_address.as_deref())
+    {
+        if let Some(name) = cur_state.interfaces.iter().find_map(|iface| {
+            if iface.base_iface().mac_address.as_deref() == Some(mac) {
+                Some(iface.name().to_string())
+            } else {
+                None
+            }
+        }) {
+            return Some(name);
+        }
+    }
+    None
```
For VLAN, MACVLAN and MACSEC, ethernet.mac-address is the MAC of the parent, not of the interface itself.

cathay4t replied:
Fixed.

ihuguet commented on rust/src/lib/nm/profile_search.rs:
```diff
@@ -0,0 +1,367 @@
+    cur_state: NetworkState,
+}
+
+impl ProfileSearch {
```
Suggestion: rename to NmConnectionMatcher.

cathay4t replied:
Changed.

ihuguet commented on rust/src/lib/nm/profile.rs:
Rename this file to `connection.rs`? All the references in the code are about `NMConnection`...

cathay4t replied:
I personally don't like `Connection` because it might confuse with other terminology on TCP or socket connection. But I don't have strong opinion on this, hence all profile is changed to connection now, both file name and variable name.

ihuguet commented:
Please try to split in smaller logical commits for the final review.

cathay4t commented:
> Please try to split in smaller logical commits for the final review.

I will split them into small parts once CI are all happy. Still coding.

cathay4t committed:
nm: Introduce `NmSettingConnection.timestamp`

Introducing `NmSettingConnection.timestamp` holding the last successful
activation time since EPOCH in seconds.

This property will be used for nmstate to choose the correct
NmConnection to modify when multiple NmConnection founds.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm: Add NmSettingConnection.autoconnect_priority

Introducing `NmSettingConnection.autoconnect_priority`, it holds `i32` for
priority when multiple NmConnection is capable for activating the same
NmDevice, the bigger number is better.

This property will be use for choosing the correct NmConnection to
modify when `nmstatectl apply` with multiple connections in NM.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm: Fix HSR support

Support handling HSR interface type when generating `BaseInterface` from
`NmDevice`.

This was not a problem because old patch ws using `NmConnection` to
build up `BaseInterface`. When we refactor the NmConnection search code,
we unified the effort by using NmDevice only for querying.

Existing HSR test cases are enough for testing this patch.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm: Use NmDevice object path of NmActiveConnection for `reapply()`

Previously we are depending `NmApi::connection_reapply()` to find the
correct NmDevice to reapply. But `test_reapply_bond_port_ref_by_mac`
test case sometimes noticed it choose the wrong NmDevice which lead to
reapply failure.

The root cause is bond ports are holding the same MAC address after
attached to a bond. If we search NmDevice by interface type and MAC, we
might get the incorrect one.

To fix this issue:

 * Expanded `NmActiveConnection` to hold NmDevice object path.
 * Changed `activate_nm_profiles()` will use `NmConnectionMatcher` which
   holds pre-apply NmActiveConnection.
 * Changed `NmApi::connection_reapply()` to take object path of
   `NmDevice` explicitly instead of guessing.

Existing test case `test_reapply_bond_port_ref_by_mac` is enough to test
this patch.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm: Fix multiconnect with `state:up` only desired state

Given a interface is activated by multiconnect NmConnection, when
applying state like

```
interfaces:
- name: eth1
  state: up
```
, nmstate will modify the multiconnect interface which is not expected.

The fix is do not return multiconnect NmConnection when
`iface.is_up_exist_config()`.

This is work well before `NmConnectionMatcher` is because:

 * Old code does not prefer activated `NmConnection` as saved profile but
   only searching name and interface type
 * No `NmConnection` been found for `iface.is_up_exist_config()`.
 * Nmstate using current state to create new `NmConnection`.

Existing test case `test_preserve_ip_settings_from_multiconnect` is
enough to test this case.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
PR is ready for review.

ihuguet commented on rust/src/lib/nm/nm_dbus/connection/iface_type.rs:
```diff
@@ -175,3 +176,58 @@ impl NmIfaceType {
+            InterfaceType::Hsr => NmIfaceType::Hsr,
+            InterfaceType::Ipsec => NmIfaceType::Vpn,
+            InterfaceType::IpVlan => NmIfaceType::Ipvlan,
+            _ => NmIfaceType::from(iface_type.to_string().as_str()),
```
Maybe we don't want to add a "catch all" here, so we don't forget to add the conversion when a new InterfaceType is added. In the NmIfaceType->InterfaceType conversion is OK because NM could add new types that we don't support yet.

cathay4t replied:
Great point. Fixed.

ihuguet commented on rust/src/lib/nm/nm_dbus/connection/iface_type.rs:
```diff
@@ -175,3 +176,58 @@ impl NmIfaceType {
     }
 }
+
+impl From<&NmIfaceType> for InterfaceType {
```
It seems this trait is actually unused?

cathay4t replied:
Used by `nm/conn_matcher.rs`:

```rust
    // Veth and ethernet will be unified to InterfaceType::Ethernet
    let iface_type = InterfaceType::from(nm_conn.iface_type()?);
```

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -0,0 +1,403 @@
+//
+// Using String instead of &str for interface name indexing because VLAN
+// connection can have auto-generated `connection.interface-name` when it is
+// empty.
```
I insist on deleting this comment, it just explain normal Rust stuff, not anything specific about nmstate logic. Actually, this is a common problem for any similar design in Rust, because indexing by `&str` immediately causes that the whole struct cannot mutate because you cannot get a `&mut`.

About `NMIfaceType`: it is mostly a C-like enum. I would consider implementing `Copy` for it, as it simplifies a lot working with it. Only `NMIfaceType::Other` can be a bit expensive to copy but it's used almost nowhere, and I wonder if we could just make it not to store a string, as it's not used for anything valuable (just log the name of the unknown type when you fail to parse and it's enough). Same for `InterfaceType`. What do you think?

cathay4t replied:
Comment removed.

Add new patch to remove `NmIfaceType::Other` and enable `Copy` derive.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -0,0 +1,403 @@
+// Using String instead of &str for interface name indexing because VLAN
+// connection can have auto-generated `connection.interface-name` when it is
+// empty.
+#[allow(dead_code)]
```
Remove this line

cathay4t replied:
Removed.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -0,0 +1,403 @@
+    acs_by_name_and_type:
+        HashMap<(String, NmIfaceType), Rc<NmActiveConnection>>,
+
+    // (name, nm_iface_type) to Rc<NmConnection>
+    // Veth will also stored into `NmIfaceType::Ethernet`
```
I'd remove these comments, the name of the fields is clear enough.

cathay4t replied:
Removed.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -0,0 +1,403 @@
+    saved_by_mac: HashMap<(String, NmIfaceType), Vec<Rc<NmConnection>>>,
+}
+
+#[allow(dead_code)]
```
Remove

cathay4t replied:
Removed.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -0,0 +1,403 @@
+
+        let mut nm_conns: Vec<&NmConnection> =
+            nm_conns.iter().map(|n| n.as_ref()).collect();
+        nm_conns.sort_unstable_by(|a, b| {
+            nm_conn_activation_sort_keys(a)
+                .cmp(&nm_conn_activation_sort_keys(b))
+        });
```
Nit: sort_unstable_by_key would be easier to read here.

cathay4t replied:
Fixed.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -0,0 +1,403 @@
+                )) {
+                    for nm_conn in nm_conns {
+                        ret.push(nm_conn);
+                    }
```
ret.extend ?

cathay4t replied:
Fixed.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -0,0 +1,403 @@
+        &self,
+        name: &str,
+        iface_type: &NmIfaceType,
+    ) -> Vec<&NmConnection> {
```
This function is used mainly to iterate the returned connections. It would make sense to return an iterator, avoiding Vec allocation, and let the caller to do `collect` if needed.

cathay4t replied:
The code will look fancy because `Box<dyn Trait>` is required for this case because map action on `Rc::ref` changed the iterator type, we need a dynamic struct implementing the Iterator trait.

Anyway, looks like this one, hope new learner of rust don't be scared.

```rust
    pub(crate) fn get_saved_by_name_type<'a>(
        &'a self,
        name: &str,
        iface_type: &NmIfaceType,
    ) -> Box<dyn Iterator<Item = &'a NmConnection> + 'a> {
        if let Some(nm_conns) = self
            .saved_by_name_and_type
            .get(&(name.to_string(), iface_type.clone()))
        {
            Box::new(nm_conns.iter().map(Rc::as_ref))
        } else {
            Box::new(std::iter::empty::<&NmConnection>())
        }
    }
```

ihuguet replied:
Ok, you are right. In this case the benefit is very small because you still need to allocate memory for the iterator, and also the code is less clean. Leave it as is or revert to the previous code, as you prefer.

cathay4t replied:
Let's keep it as `no-modify` sign. :)

ihuguet commented on rust/src/lib/nm/show.rs:
```diff
@@ -80,64 +76,40 @@ pub(crate) async fn nm_retrieve(
-                } else {
-                    None
-                };
+        if iface.is_ignore() {
```
This doesn't seem to be exactly the same as before. Before it was checked if `nm_dev.state` is `Unmanaged` or `Disconnected`. With `is_ignore` you only check if it is `Unmanaged` because `nm_dev_to_nm_iface` only sets `Ignore` in that case.

cathay4t replied:
Previously code on ignoring `nm_dev.state == Disconnected` is causing https://issues.redhat.com/browse/RHEL-86240 . This is change is required for fixing that issue.

ihuguet replied:
OK, got it. Please try to split this kind of changes in a separate commit with a proper explanation (next time). For example, first refactor to use conn_matcher but maintaining previous behaviour, and then other commit with the fix.

ihuguet commented on rust/src/lib/net_state.rs:
```diff
@@ -345,6 +345,7 @@ pub(crate) struct MergedNetworkState {
     pub(crate) rules: MergedRouteRules,
     pub(crate) memory_only: bool,
     pub(crate) override_iface: bool,
+    pub(crate) current: NetworkState,
```
Is this used for something?

cathay4t replied:
It is required for `get_nm_connection_iface_name` in `NmConnectionMatcher`.

I moved this change into commit `nm: Use NmConnectionMatcher for apply action` where it is required.

ihuguet replied:
Was it not possible to get this info from `MergedNetworkState::interfaces` and then `MergedInterface::current`?

Just asking, I was surprised that the whole NetworkState is stored for such a small thing, moreover inside of MergedNetworkState which already has the information of the current network state.

cathay4t replied:
It is possible reading current interface from MergedNetworkState . Let me polish the code.

cathay4t replied:
Code updated and taking `MergedNetworkState` only.

ihuguet replied:
Great! Thanks!

ihuguet commented on rust/src/lib/nm/query_apply/apply.rs:
```diff
@@ -455,23 +401,21 @@ fn gen_nm_conn_need_to_deactivate_first(
-                        false
-                    }
-                })
+            // VPN connection does not applied NmConnection
```
Incomplete comment?

cathay4t replied:
Fixed.  Changed to `            // VPN connection does not have applied NmConnection`

ihuguet commented on rust/src/lib/nm/query_apply/profile.rs:
```diff
@@ -1,80 +1,46 @@
+    {
+        for saved_nm_conn in conn_matcher
+            .get_saved(merged_iface.merged.base_iface())
+            .into_iter()
+            .filter(|nm_conn| {
+                nm_conn.flags.contains(&NmSettingsConnectionFlag::Volatile)
+            })
```
Before the changes we were excluding volatile connections. Now we exclude all but volatile connections. I guess it is a mistake.

cathay4t replied:
Nice catch! Fixed.

ihuguet commented on rust/src/lib/nm/query_apply/profile.rs:
```diff
@@ -1,80 +1,46 @@
     nm_api: &mut NmApi<'_>,
-    exist_nm_conns: &[NmConnection],
-    nm_conns: &[NmConnection],
+    merged_state: &MergedNetworkState,
```
I'm not very convinced of having to pass the entire MergedNetworkState. This is a function with a very specific task, so it's better if it just require the really minimum info that it needs to do its work. Otherwise, we'll increase coupling between different components more each time, not less.
Is there anything we can do to simplify it? Maybe require a `&[&BaseInterface]`?

cathay4t replied:
For this `&[&BaseInterface]`, you have to create `Vec<&BaseInterface>` to own it, passing the entire `&MergedNetworkState` does not require any memory allocation.

Passing iterator of `&BaseInterface` also require memory allocation on iterator itself, the code might require `Box<dyn trait>` as I mentioned above, seem overkill to me.

ihuguet replied:
OK, I see that others are accepting MergedInterfaces, like delete_orphan_ovs_ports, so I won't insist on this.

ihuguet replied:
BTW, for function arguments you can use `&dyn Trait` without a `Box`. On return data is more difficult and normally not possible.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -237,15 +237,27 @@ impl NmConnectionMatcher {
         &self,
         base_iface: &BaseInterface,
     ) -> Vec<&NmConnection> {
-        let mut ret: Vec<&NmConnection> = if let Some(nm_conns) =
-            self.saved_by_name_and_type.get(&(
-                base_iface.name.to_string(),
-                NmIfaceType::from(&base_iface.iface_type),
-            )) {
-            nm_conns.iter().map(Rc::as_ref).collect()
-        } else {
-            Vec::new()
-        };
+        let mut ret: Vec<&NmConnection> = Vec::new();
+        // If interface type is unknown, all profiles with desired interface
+        // name will be returned.
+        if base_iface.iface_type == InterfaceType::Unknown {
+            for nm_conns in self.saved_by_name_and_type.iter().filter_map(
+                |((name, _), nm_conns)| {
+                    if name == base_iface.name.as_str() {
+                        Some(nm_conns)
+                    } else {
+                        None
+                    }
+                },
+            ) {
+                ret.extend(nm_conns.iter().map(Rc::as_ref));
+            }
```
IMO it would make sense to add a new index `saved_by_name` at creation time, and use it here.

cathay4t replied:
Fixed.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -237,15 +237,27 @@ impl NmConnectionMatcher {
-            Vec::new()
-        };
+        let mut ret: Vec<&NmConnection> = Vec::new();
+        // If interface type is unknown, all profiles with desired interface
+        // name will be returned.
```
Please document this above the function, in its doc-comment.

cathay4t replied:
Fixed.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -232,6 +233,53 @@ impl NmConnectionMatcher {
+    ) -> Option<&'s NmConnection> {
+        let nm_iface_type = NmIfaceType::from(&base_iface.iface_type);
+
+        if let Some(nm_conn) = self
+            .acs_by_name_and_type
+            .get(&(base_iface.name.to_string(), nm_iface_type.clone()))
+            .and_then(|nm_ac| self.saved_by_uuid.get(nm_ac.uuid.as_str()))
+            .map(Rc::as_ref)
+        {
+            return Some(nm_conn);
+        }
```
This block seems unnecessary, and maybe wrong. If user requests match by MAC, this block might get a match by ifname first. I'm not really sure how that can actually happen, I guess that a different NIC has to be called `port1`, which is quite confusing, at the very least, or maybe directly wrong user's desired state. But anyway, I think this block should be removed.

cathay4t replied:
This is searching on NmActiveConnection by name and interface, I think it is reasonable to use activated NmConnection when modifying.

ihuguet replied:
Please add a comment explaining this

cathay4t replied:
Command added:

```
        // Prefer the current activated NmConnection by searching
        // NmActiveConnection for specified interface, then use the UUID of
        // NmActiveConnection to search out the NmConnection.

```

ihuguet commented on rust/src/lib/nm/nm_dbus/active_connection.rs:
```diff
@@ -17,6 +17,7 @@ pub struct NmActiveConnection {
     pub iface_type: NmIfaceType,
     pub iface_name: String,
     pub state_flags: u32,
+    pub dev_obj_path: Option<String>,
```
Is it really possible to have an ActiveConnection without a device? Otherwise, better not to use Option.

cathay4t replied:
No idea. But this layer is just a wrapper of DBUS.

cathay4t committed:
nm: implement From trait between InterfaceType and NmIfaceType

Previously, the `nm_dbus` was designed to be standalone crate without
dependency on nmstate, so `iface_type_to_nm()` is used to convert from
`InterfaceType` to `NmIfaceType`. But we don't have plan to release
`nm_dbus` crate anymore, so we should use From trait instead of function
for conversion between `InterfaceType` and `NmIfaceType`.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm: Delete deactivated NM connection also

With a dummy NM connection created by deactivated, applying desire state
like:

```
interfaces:
  - name: dummy1
    state: absent
```

will not delete the NM connection.

This is because `NmConnectionMatcher::get_saved()` is searching for
NmConnection with `NmIfaceType::Unknown`.

To fix that issue, when `get_saved()` got unknown interface, all
`NmConnection` matching specified name will be returned.

The integration test case `test_state_absent_can_remove_down_profiles`
can reproduce the issue and confirm fix.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm: Rename profile.rs to connection.rs

Since we are using NmConnection, it is reasonable to call its holder
file as `connection.rs` instead of `profile.rs`.

Also renamed all reference on `NmConnection` from profile to connection.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm: Fix NmConnectionMatcher for modifying MAC based interface

When modifying existing NmConnection with desired state:

```
---
interfaces:
- name: port1
  type: ethernet
  mac-address: 00:23:45:67:89:1a
  identifier: mac-address
```
The `NmConnectionMatcher::get_prefered_saved()` will return empty
None because name in this case is `port1` instead of the actual NIC
name.

Fixed by rename existing `NmConnectionMatcher::get_prefered_saved()` to
`get_prefered_saved_by_name_type()` and the new
`NmConnectionMatcher::get_prefered_saved()` is accepting
`BaseInterface`.

When got `InterfaceIdentifier::MacAddress`, we search the MAC address
and interface type instead of its profile name.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
Can delete non-exist mac-ref connection

When a NM connection is porting to non-exist(or deleted) ethernet
interface, `state: absent` will not works because top level reject this
due to interface not exist and also discard the `identifier` and
`mac-address` during deserialization.

Fixed by:
 1. Preserve `mac-address`, `identifier` during deserialization for
    `state: absent` interface`.
 2. Skip resolving absent interface's ref-mac.

Integration test case included.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm: Remove NmIfaceType::Other

The `NmIfaceType::Other` is rarely used, removing it would able to
enable `Copy` derive on `NmIfaceType`.

Signed-off-by: Gris Ge <fge@redhat.com>

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -0,0 +1,468 @@
+                self.saved_by_mac
+                    .entry((mac.to_uppercase(), NmIfaceType::Ethernet))
+                    .or_default()
+                    .push(nm_conn.clone())
```
This is inverted. It should be 
```
if nm_iface_type == Veth {
    insert(..., Ethernet)
}
insert(..., nm_iface_type)
```

cathay4t replied:
Fixed.

ihuguet commented on rust/src/lib/nm/conn_matcher.rs:
```diff
@@ -0,0 +1,468 @@
+            self.saved_by_mac
+                .entry((mac.to_uppercase(), *nm_iface_type))
+                .or_default()
+                .push(nm_conn.clone())
```
Does not this one need to be removed? It might insert empty MACs. It will also insert an entry for VLANs and duplicated entry for non-VLAN (not really inserted because it already exists, but anyway, not needed).

cathay4t replied:
Fixed.

cathay4t committed:
nm: Unify the effort on searching NmConnection

Creating struct `NmConnectionMatcher` to unify the effort on searching
NmConnection for querying and modification.

The `NmConnectionMatcher` will create indexes at initialization to speed
up follow up queries.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm: Use `NmConnectionMatcher` for querying and apply

Use `NmConnectionMatcher` for matching NmDevice with NmConnection during
`nmstatectl show` and `nmstatectl apply`.

Resolves: https://issues.redhat.com/browse/RHEL-92916
Resolves: https://issues.redhat.com/browse/RHEL-86240

Signed-off-by: Gris Ge <fge@redhat.com>
